File: golang/binary-tree-bfs/avg_of_level_binary_tree.go
package main

func averageOfLevels(root *TreeNode) []float64 {
	if root == nil {
		return []float64{}
	}

	averages := []float64{}
	queue := []*TreeNode{root}

	for len(queue) > 0 {
		levelSize := len(queue)

		levelSum := 0

		for i := 0; i < levelSize; i++ {
			node := queue[0]
			queue = queue[1:]

			levelSum += node.Val

			if node.Left != nil {
				queue = append(queue, node.Left)
			}

			if node.Right != nil {
				queue = append(queue, node.Right)
			}
		}

		averages = append(averages, float64(levelSum)/float64(levelSize))
	}

	return averages
}

File: java/src/main/java/com/sanedge/interview/binary_tree_bfs/AvgOfLevelBinaryTree.java
package com.sanedge.interview.binary_tree_bfs;

import com.sanedge.interview.utils.TreeNode;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class AvgOfLevelBinaryTree {

    public List<Double> averageOfLevels(TreeNode root) {
        List<Double> averages = new ArrayList<>();

        if (root == null) {
            return averages;
        }

        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            int levelSize = queue.size();
            double levelSum = 0;

            for (int i = 0; i < levelSize; i++) {
                TreeNode node = queue.poll();

                levelSum += node.val;

                if (node.left != null) {
                    queue.offer(node.left);
                }

                if (node.right != null) {
                    queue.offer(node.right);
                }
            }

            averages.add(levelSum / levelSize);
        }

        return averages;
    }
}
